// @reservoir0x/sdk/dist/wyvern-v2/builders/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builders = void 0;
const erc721_1 = require("./contract-wide/erc721");
const erc1155_1 = require("./contract-wide/erc1155");
const erc721_2 = require("./single-token/erc721");
const erc1155_2 = require("./single-token/erc1155");
const erc721_3 = require("./token-list/erc721");
const erc1155_3 = require("./token-list/erc1155");
const erc721_4 = require("./token-range/erc721");
const erc1155_4 = require("./token-range/erc1155");
exports.Builders = {
    Erc721: {
        ContractWide: erc721_1.ContractWideErc721Builder,
        SingleToken: erc721_2.SingleTokenErc721Builder,
        TokenList: erc721_3.TokenListErc721Builder,
        TokenRange: erc721_4.TokenRangeErc721Builder,
    },
    Erc1155: {
        ContractWide: erc1155_1.ContractWideErc1155Builder,
        SingleToken: erc1155_2.SingleTokenErc1155Builder,
        TokenList: erc1155_3.TokenListErc1155Builder,
        TokenRange: erc1155_4.TokenRangeErc1155Builder,
    },
};
