// @reservoir0x/sdk/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.s = exports.n = exports.lc = exports.getCurrentTimestamp = exports.bn = exports.getRandomBytes32 = exports.MaxUint256 = exports.BytesEmpty = void 0;
const bignumber_1 = require("@ethersproject/bignumber");
const random_1 = require("@ethersproject/random");
// Constants
exports.BytesEmpty = "0x";
exports.MaxUint256 = bignumber_1.BigNumber.from("0x" + "f".repeat(64));
// Random
const getRandomBytes32 = () => (0, exports.bn)((0, random_1.randomBytes)(32));
exports.getRandomBytes32 = getRandomBytes32;
// BigNumber
const bn = (value) => bignumber_1.BigNumber.from(value);
exports.bn = bn;
// Time
const getCurrentTimestamp = (delay = 0) => Math.floor(Date.now() / 1000 + delay);
exports.getCurrentTimestamp = getCurrentTimestamp;
// Ease of use
const lc = (x) => x === null || x === void 0 ? void 0 : x.toLowerCase();
exports.lc = lc;
const n = (x) => Number(x);
exports.n = n;
const s = (x) => String(x);
exports.s = s;
