// @ramosdiego/ui/dist/buttons.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const userStyles_1 = __importDefault(require("./userStyles"));
function buttons(options, theme, addBase, colors) {
    const components = {};
    const { buttons, globalStyles } = options;
    const userGlobalStyles = (0, userStyles_1.default)(theme, globalStyles);
    const userButtonBaseStyles = (0, userStyles_1.default)(theme, buttons === null || buttons === void 0 ? void 0 : buttons.baseStyles);
    const presets = {
        elegant: {
            borderRadius: theme(`borderRadius[none]`),
            fontWeight: theme(`fontWeight[light]`),
            textTransform: 'uppercase',
            borderWidth: '1px',
        },
        playful: {
            borderRadius: theme(`borderRadius[full]`),
            fontWeight: theme(`fontWeight[semibold]`),
            textTransform: 'normal-case',
        },
    };
    let preset = {
        borderRadius: theme(`borderRadius[DEFAULT]`),
        fontWeight: theme(`fontWeight[medium]`),
        textTransform: 'uppercase',
    };
    if ((buttons === null || buttons === void 0 ? void 0 : buttons.preset) && presets[buttons === null || buttons === void 0 ? void 0 : buttons.preset]) {
        preset = presets[buttons === null || buttons === void 0 ? void 0 : buttons.preset];
    }
    const buttonBaseStyles = Object.assign(Object.assign(Object.assign({ padding: `${theme(`spacing[2]`)} ${theme(`spacing[4]`)}`, borderWidth: theme(`borderWidth[2]`), borderColor: theme('borderColor[transparent]'), userSelect: 'none', outline: '2px solid transparent', outlineOffset: '2px', '&:disabled': {
            opacity: '80%',
            cursor: 'not-allowed',
        }, transitionProperty: theme(`transitionProperty[all]`), transitionDuration: theme(`transitionDuration[200]`), transitionTimingFunction: theme(`transitionTimingFunction['ease-in-out']`), display: 'flex', alignItems: 'center', justifyContent: 'center', gap: theme(`spacing[2]`) }, preset), userGlobalStyles), userButtonBaseStyles);
    if (buttons === null || buttons === void 0 ? void 0 : buttons.animate) {
        addBase({
            '@keyframes push': {
                from: {
                    transform: 'translateY(0%)',
                },
                to: {
                    transform: 'translateY(5%)',
                },
            },
        });
        buttonBaseStyles['&:active'] = {
            animation: 'push 0.09s ease-in-out',
        };
    }
    colors.forEach((color) => {
        components[`.btn-${color}-fill`] = Object.assign({ backgroundColor: theme(`colors.${color}[100]`), color: theme(`colors.${color}[900]`), '&:not(:disabled):hover': {
                backgroundColor: theme(`colors.${color}[50]`),
            }, '&:focus': {
                backgroundColor: theme(`colors.${color}[50]`),
                boxShadow: `0 0 0 2px ${theme(`colors.${color}[300]`)}`,
            }, '&:disabled': Object.assign(Object.assign({}, buttonBaseStyles['&:disabled']), { color: theme(`colors.neutral[600]`), backgroundColor: theme(`colors.neutral[100]`) }) }, buttonBaseStyles);
        components[`.btn-${color}-fill-dark`] = Object.assign({ backgroundColor: theme(`colors.${color}[900]`), color: theme(`colors.${color}[50]`), '&:not(:disabled):hover': {
                color: theme(`colors.white`),
                backgroundColor: theme(`colors.${color}[800]`),
            }, '&:focus': {
                boxShadow: `0 0 0 2px ${theme(`colors.${color}[500]`)}`,
            }, '&:disabled': Object.assign(Object.assign({}, buttonBaseStyles['&:disabled']), { color: theme(`colors.neutral[50]`), backgroundColor: theme(`colors.neutral[600]`) }) }, buttonBaseStyles);
        components[`.btn-${color}-outline`] = Object.assign({ borderColor: theme(`colors.${color}[200]`), color: theme(`colors.${color}[900]`), '&:not(:disabled):hover': {
                backgroundColor: theme(`colors.${color}[50]`),
            }, '&:focus': {
                backgroundColor: theme(`colors.${color}[50]`),
                boxShadow: `0 0 0 2px ${theme(`colors.${color}[300]`)}`,
            }, '&:disabled': Object.assign(Object.assign({}, buttonBaseStyles['&:disabled']), { color: theme(`colors.neutral[600]`), borderColor: theme(`colors.neutral[200]`) }) }, buttonBaseStyles);
        components[`.btn-${color}-outline-dark`] = Object.assign({ borderColor: theme(`colors.${color}[800]`), color: theme(`colors.${color}[50]`), '&:not(:disabled):hover': {
                color: theme(`colors.white`),
                backgroundColor: theme(`colors.${color}[800]`),
            }, '&:focus': {
                color: theme(`colors.white`),
                backgroundColor: theme(`colors.${color}[800]`),
                boxShadow: `0 0 0 2px ${theme(`colors.${color}[500]`)}`,
            }, '&:disabled': Object.assign(Object.assign({}, buttonBaseStyles['&:disabled']), { color: theme(`colors.neutral[50]`), borderColor: theme(`colors.neutral[600]`) }) }, buttonBaseStyles);
        components[`.btn-${color}-ghost`] = Object.assign({ color: theme(`colors.${color}[900]`), '&:not(:disabled):hover': {
                backgroundColor: theme(`colors.${color}[50]`),
            }, '&:focus': {
                backgroundColor: theme(`colors.${color}[50]`),
                boxShadow: `0 0 0 2px ${theme(`colors.${color}[300]`)}`,
            }, '&:disabled': Object.assign(Object.assign({}, buttonBaseStyles['&:disabled']), { color: theme(`colors.neutral[600]`) }) }, buttonBaseStyles);
        components[`.btn-${color}-ghost-dark`] = Object.assign({ color: theme(`colors.${color}[50]`), '&:not(:disabled):hover': {
                color: theme(`colors.white`),
                backgroundColor: theme(`colors.${color}[800]`),
            }, '&:focus': {
                color: theme(`colors.white`),
                backgroundColor: theme(`colors.${color}[800]`),
                boxShadow: `0 0 0 2px ${theme(`colors.${color}[500]`)}`,
            }, '&:disabled': Object.assign(Object.assign({}, buttonBaseStyles['&:disabled']), { color: theme(`colors.neutral[50]`), backgroundColor: theme(`colors.nuetral[900]`) }) }, buttonBaseStyles);
    });
    return components;
}
exports.default = buttons;
